### ISSUE TITLE
Idea for a more useful variant of `Hint Mode` `+`

### ISSUE THREAD

**Issue by samuelgruetter** [coq/coq#18078]
Today I learnt that passing a `+` to `Hint Mode` does not actually do what I thought it does. It means that typeclass search should treat the argument at the corresponding position as an input, so I used to think that this means that typeclass search will refrain from instantiating any evars appearing in that argument.
But in fact, the `+` mode restricts typeclass search in a much stronger way: It completely prevents typeclass search from running if the argument contains any evar.

The [documentation](https://coq.inria.fr/refman/proofs/automatic-tactics/auto.html#coq:cmd.Hint-Mode) is quite clear about the meaning of `+`:

> mode `+` matches a term if and only if it does not contain existential variables

so there is no bug here, just an idea for a potentially useful feature:

How about adding a new mode, eg `i` (or any other character that seems appropriate), whose meaning is that the evars appearing in the argument at this position may not be instantiated by typeclass search, unless they also appear in an argument marked as output (ie anywhere in a `-` argument, or at a non-head position of a `!` argument).

Below is an example where this feature might be useful: The second argument of `safe_implication` needs to be treated as an input, otherwise the wrong hint will be used, but if I use such a `Hint Mode`, typeclass search doesn't run any more at all, and I have to use `hide_evars` before to make sure it runs.

```coq
(* Meaning: P implies Q and the only reasonable way to
   prove Q is to try proving P. *)
Definition safe_implication(P: Prop)(Q: Prop): Prop := P -> Q.

Create HintDb safe_implication.

Ltac apply_safe_implication :=
  lazymatch goal with
  | |- ?Q =>
      let H := fresh in
      eassert (safe_implication _ Q) as H by (typeclasses eauto with safe_implication);
      unfold safe_implication in H;
      apply H;
      clear H
  end.

Parameter f: nat -> nat -> Prop.
Parameter g: nat -> Prop.
Parameter h: nat -> Prop.

Axiom f_same_args: forall x, safe_implication (g x) (f x x).
Axiom f_first_arg_0: forall x, safe_implication (h x) (f 0 x).

#[export] Hint Resolve f_first_arg_0 f_same_args : safe_implication.

#[global] Hint Mode safe_implication - + : safe_implication.

Ltac hide_evars :=
  repeat match goal with
    | |- context[?x] => is_evar x; let y := fresh "e0" in set (y := x)
    end.

Ltac unhide_evars :=
  repeat match goal with
    | e := ?rhs |- _ => is_evar rhs; subst e
    end.

Goal forall a, 0 < a -> h a -> exists b, f 0 b /\ 0 < b.
Proof.
  intros. eexists. split.

  hide_evars.
  apply_safe_implication.
  unhide_evars.
  all: eassumption.
Qed.
```

**Comment by Janno** [comment]
I like this idea a lot. I always wanted a mode that says "tolerate but don't instantiate evars in this argument". 

 I wonder if `Typeclasses Strict Resolution` (https://coq.inria.fr/refman/addendum/type-classes.html#coq:flag.Typeclasses-Strict-Resolution) should be repurposed to do this. Right now it does not interact usefully with modes, making it entirely useless for classes that have outputs.

**Comment by RalfJung** [comment]
Yeah that sounds great! It could probably replace our use of mode `!` almost everywhere we use it in Iris.
Cc @robbertkrebbers

**Comment by robbertkrebbers** [comment]
I completely agree with this proposal.

From discussions with @gares I gathered that your proposal is exactly what input modes in Prolog, and Elpi in particular, do.

**Comment by SkySkimmer** [comment]
Branch with experimental implementation https://github.com/coq/coq/pull/18099

It makes `+` behave as the proposed `i` for typeclasses.

When there are modes the Strict Resolution option is ignored.

It is broken with eauto (`+` gets ignored).

Perf is probably relatively not optimized.

**Comment by gares** [comment]
> From discussions with @gares I gathered that your proposal is exactly what input modes in Prolog, and Elpi in particular, do.

I don't know about Prolog in general, but that is definitely what the input mode of Elpi does. The argument in input mode is matched (and not unified) against the pattern coming from the rule. Variable in the rule's pattern can be instantiated, variables in the input term are not (the input term is untouched).

**Comment by gares** [comment]
> Branch with experimental implementation #18099

@SkySkimmer I believe the implementation is not super trivial in Coq, since a term like `C ?1 ?1` against an instance pattern `C _ x` with mode `C i _` should "unify". So `?1` is locked only when it occurs in the first argument, but not when it occurs in the second.

**Comment by SkySkimmer** [comment]
It does unify in my implementation 
However it also unifies against `C x _` with mode `C i _` which is maybe not wanted

**Comment by gares** [comment]
No, that should fail, since it would make ?1 := x.

**Comment by samuelgruetter** [comment]
When I wrote this proposal, I only considered specifications that fit into the following scheme:
* First, by looking at the goal and at the mode, determine for each evar whether it's ok to instantiate it or not (and I guess those that must not be instantiated are called "locked", if I got the terminology right)
* Then, run typeclass search, and prevent it from instantiating any locked evars while doing unifications, but also while running arbitrary Ltac code that was registered with `Hint Extern`.

So, first, do we all agree that the feature we're aiming for fits into the above scheme? (I'm asking especially since the comment by @gares [above](https://github.com/coq/coq/issues/18078#issuecomment-1740869170) might disagree).

But even within this scheme, I see multiple reasonable specifications of what such a new mode should do:

For evars that only appear at input positions, it's clear that they should be locked (not instantiatable), and for evars that only appear at output positions, it's clear that they should not be locked (ie instantiatable), but:
* [both] What about evars that appear at input as well as output positions?
* [neither] What about evars that don't appear at any input or output positions (but might get instantiated by some Ltac code that was registered with `Hint Extern`)?

Originally, I meant to suggest that all evars are considered unlocked by default, then all evars at input positions get locked, and then, those also appearing at output positions get unlocked again, i.e. in short: "both:unlocked, neither:unlocked".
But now it seems that actually, all 2^2 flavors could make sense, i.e. I could also imagine use cases for "both:unlocked, neither:locked", "both:locked, neither:unlocked" and "both:locked, neither:locked". So does that mean we need four different flavors of this new feature? Or, as @Janno already suggested, an integration of `Hint Mode` and `Typeclasses Strict Resolution`?

**Comment by samuelgruetter** [comment]
Instead of locking whole *evars*, it might make more sense to lock specific *evar occurrences*, as @andres-erbsen pointed out to me. Then, if an evar appears both in an input and output position, its occurrence in the input would be locked, but not its occurrence in the output. However, that doesn't seem quite compatible with the [current AllowedEvars mechanism](https://github.com/coq/coq/blob/b5d259a7a5d13eb65524cb84f8c92c187bcb1a32/pretyping/evarsolve.ml#L28-L51).
And in cases where typeclass search applies a lemma of type `P y -> Q y y`, and the goal is `Q ?x ?x`, with the first occurrence of `?x` locked and the second unlocked, one would have to decide whether unifying a locked occurrence with an unlocked one results in a locked or unlocked evar in the remaining `P ?x` goal.

**Comment by SkySkimmer** [comment]
>one would have to decide whether unifying a locked occurrence with an unlocked one results in a locked or unlocked evar in the remaining P ?x goal.

Don't you just look at the mode of P?

**Comment by samuelgruetter** [comment]
> Don't you just look at the mode of P?

ah yes, that makes sense :+1:

**Comment by Blaisorblade** [comment]
> it might make more sense to lock specific _evar occurrences_,

That seems to match what Enrico advocated in https://github.com/coq/coq/issues/18078#issuecomment-1740869170 ? Just wondering if you missed the comment and there's consensus, or if there's an actual difference.

**Comment by gares** [comment]
What I had in mind is not to lock evars (or some occurrences) once and for all, but rather reconsinder their status at each goal, checking the mode of the goal. So if you have a rule (an instance with a parameter) of type `forall x, P x -> Q x`, then
the goal `Q ?x` may see ?x as locked, but the subsequent `P ?x` may be allowed to instantiate `?x` if its mode is more liberal.
So I think we are on the same page w.r.t. the spec. But I'm not so sure the implementation is that simple.

The tricky case I was pointing out to Gaetan is where the same evar occurs in two arguments of different mode. In this case I believe "locking" the evar is hard since one needs to change its locking status during an unification procedure, and I don't think Coq has any support for this, as @samuelgruetter points out in the linked code snippet.

Not to talk about that unification can postpone sub problems (and the predicate may have been crossed already by congruence), so one would need to save the locked attribute of evars together with the suspended sub problem in order to restore it when the problem gets resumed.

IMO it is simpler to have the entire unification procedure take an extra boolean `~matching_mode` and, in that case, disable all branches that would instantiate, say, the LHS. Then if the goal is `P a b c` and the rule's head `P a' b' c'`, one
calls unification 3 times with a value of `~matching_mode` which depends on the mode declared by P, and that mode won't change for the whole call to unification (or matching actually). This is how it is done in Elpi, unification variables don't have any attribute to lock them. And the order in which you unify these arguments is part of the spec (if the same unif variable occurs in both input and output arguments, then matching may fail if you go L2R, but succeed if you R2L for example, I can craft an example).

**Comment by robbertkrebbers** [comment]
Just curious, are there any updates about this? Is it likely that this would happen at some point in Coq, or only once we get @gares's and @FissoreD's new TC engine, because it's impossible with the current implementation?

The reason I ask is that we recently had some discussions about it in Iris, https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/642#note_103628

**Comment by Janno** [comment]
I think I understand the problem pointed out in this discussion but I am a little worried that the concerns are blocking a useful feature that could be quite easy to implement and would fit at least 80% of all the use cases. Freezing all evars in an argument with mode `=` for the entire unification call will work for every single use case I have ever had for this feature. The absence of concrete examples in this discussion makes me wonder if it wouldn't perhaps cover most other people's use cases as well.

The use cases I have in mind are all very boring and often involve classes that express properties of terms, usually with fields that carry proofs of these properties. For example, in the context of iris, a separation logic assertion can be `Persistent`. There are many instances for operators polymorphic in the concrete logic (`bi`) of the form `some_operator_persistent : forall (PROP:bi) .., @Persistent PROP (some_operator ...)`. 

Since iris is written in what I would call defensive Coq style (which I appreciate tremendously!) the modes on `Persistent` are `+ !`. The actual intention of these modes are to never instantiate evars in the query. `+ +` would do that even better but there are many operators that can be safely found to be persistent even when some of their arguments are evars. So, AFAIU, `+ !` represents a compromise that will avoid completely crazy solutions (i.e. picking the wrong `bi` if that happens to be an evar) but allow some evars to be present. However, this setup comes with the usual problem of some false positives. For example, iris's `big_op`s had instances of the form `@Persistent ?PROP (big_op bi_sep [])` that would succeed on queries of the form `@Persistent ?PROP (big_op big_sep ?ls)` and break proofs.

Still, the setup works for most use cases. However it also makes it impossible to run queries such as  `Persistent ?PROP (@bi_emp ?PROP)` despite there being an instance of exactly that type. I do not see any reason to disallow that query since it is perfectly safe. But  there are no hint modes that express the intended constraint on the `Persistent` typeclass. In cases like these, the proposed mode `i` (which I would call `=`) would work perfectly. I don't even think it is possible to write a useful instance of `Persistent` that could run into the issue that is blocking this feature. An instance whose second argument forces the instantiation of the first would automatically have a type that would force the instantiation of the first argument even without the help of the second argument. I would love to see counterexamples.

> [..] a term like `C ?1 ?1` against an instance pattern `C _ x` with mode `C i _` should "unify"

I think I disagree with this statement. I might be missing something but it seems to me that the only way for such an instance to successfully apply (disregarding modes entirely) to the goal `C ?1 ?1` is if its type is either directly convertible to `C x x` or if its type can be instantiated to be convertible to `C x x`, possibly by inferring the first argument based on the value of the second through canonical structures or other tricks. I find both scenarios to intuitively be in violation of the mode `C i _`.

Are there more concrete examples of variable sharing where the unification of output arguments require (subterms of) input arguments to be unified/instantiated in interesting ways?

**Comment by Janno** [comment]
I've been thinking about modes again and I am wondering if people have examples that actually require `Hint Mode +` as opposed to the proposed mode `i`. I personally cannot think of a single class in our code base that wants the semantics of `+` except for a gadget class called `NoEvar` which uses the mode to implement the evar checking.  It seems to me that `+` is simply the wrong mode for almost any use case I can think of. Is anybody aware of counterexamples?

**Comment by robbertkrebbers** [comment]
Assuming we would have the `i` mode you propose, I am not sure Iris would have a need for `+` or `!`. I think your assessment about the use of `+` in Iris is right: it uses that to be more defense since `!` is too permissive.

**Comment by samuelgruetter** [comment]
If we had `i`, the only use case for `+` I could imagine would be to provide backwards compatibility with brittle proofs that rely on typeclass searches involving `+` not succeeding and therefore have additional manual Ltac that would become unnecessary if `i` was used.

**Comment by andres-erbsen** [comment]
It also seems to me that `+` is just wrong for all use cases I can think; I hope the proposed `i` is indeed more suitable. It's just hard to be sure because there might be other similarly surprising typeclass-search behaviors that interact with this choice.

**Comment by robbertkrebbers** [comment]
> It's just hard to be sure because there might be other similarly surprising typeclass-search behaviors that interact with this choice.

I agree.

How hard would it be to make a prototype of the `i` proposal? I would be happy to test it on std++ and Iris and see whether we run into any of these surprising behaviors?

**Comment by Janno** [comment]
> How hard would it be to make a prototype of the `i` proposal? I would be happy to test it on std++ and Iris and see whether we run into any of these surprising behaviors?

I have created a PR (https://github.com/rocq-prover/rocq/pull/21911) and would appreciate feedback on it. Note that it does not replace `+` but instead introduces a new mode (`=`). The PR description hopefully contains all relevant information on the choices I made.
